### PR TITLE
Add GFPN class

### PR DESCRIPTION
**Pull request by vrancurel**
- Poly<T> MUST apply to sub_fields of extension fields (not the RING of integers modulo N neither an extension field)
- Add gfpn class with
- Add polynomial gcd and extended euclidean algorithm e.g. to compute the inverse of a polynomial in a field extension
- fix various polynomial functions (e.g. set() when the coef is zero, conversion functions from/to number reprentation)
- add simple extension field tests

### CODE REVIEW

File: test/gf_utest.cpp [hunk 1]
```diff
@@ -139,6 +182,40 @@ class GFUtest
...
+    test_log(&gf9);
+  }
+
+  void test_prime_root_gf9()
+  {
+    std::cout << "test_prime_root_gf9\n";
+
+    GFPN<T> gf9(3, 2);
+    test_find_prime_root(&gf9);
+    //XXX test_get_order(&gf9);
```
Comment (vrancurel) [inline test/gf_utest.cpp]: @lamphamsy , the get_order() function is probably not well adapted for GF_p^n fields, can you have a look / see how to generalize it ?

**Comment by vrancurel** [comment]
I think now it is possible to write a fecgfpnfftrs for any extension field (e.g. GF_27 !)

File: test/gf_utest.cpp (hunk 1 above)
Comment (lamphamsy) [reply test/gf_utest.cpp]: okay, I will do it

Comment (lamphamsy) [reply test/gf_utest.cpp]: it's done

File: src/gfpn.h [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  return _b2.to_num();
+}
+
+template <typename T>
+T GFPN<T>::exp(T a, T b)
+{
+  assert(this->check(a));
+  assert(this->check(b));
+
+  return GF<T>::exp_naive(a, b);
```
Comment (lamphamsy) [inline src/gfpn.h]: why don't we use exp_quick here?

Comment (vrancurel) [reply src/gfpn.h]: feel free to change !

File: src/gfpn.h [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  assert(this->check(a));
+  assert(this->check(b));
+
+  Poly<T> _a(sub_field);
+  _a.from_num(a, max_deg);
+  Poly<T> _b(sub_field);
+  _b.from_num(b, max_deg);
+
+  _a.add(&_b);
+  _a.mod(prim_poly);
```
[the remark this answers is not in the document]
Comment (vrancurel) [reply src/gfpn.h]: please remove

File: src/rn.h [hunk 4]
```diff
@@ -522,11 +522,12 @@ template <typename T>
...
-  std::vector<T> primes;
-  std::vector<T> exponent;
   T h = this->card_minus_one();
-  _factor_prime<T>(h, &primes, &exponent);
-  T order = do_step_get_order(x, h, &primes, &exponent);
+  if (!this->compute_factors_of_order_done)
+    _factor_prime<T>(h, this->primes, this->exponent);
+  std::vector<T> _primes(*primes);
+  std::vector<T> _exponent(*exponent);
+  T order = do_step_get_order(x, h, &_primes, &_exponent);
```
Comment (vrancurel) [inline src/rn.h]: @lamphamsy do we need exponent and order as T ? can't just be integers ? (maybe stupid question and we need it)

Comment (lamphamsy) [reply src/rn.h]: Since order could be (q-1), it should be as T. For exponent, I think it could be integers :)

**Pull request merged**